Port contents of `Documentation/PackageSecurity.md` to DocC (#8726)
swiftlang/swift-package-manager: Sources/PackageManagerDocs/Documentation.docc/PackageSecurity.md

# Package Security

Learn about the security features that SwiftPM implements.

## Trust on First Use

> SwiftPM records **fingerprints** of downloaded package versions so that

Reviewer: ```suggestion
Package manager records **fingerprints** of downloaded package versions so that
```

it can perform [trust-on-first-use](https://en.wikipedia.org/wiki/Trust_on_first_use)
(TOFU). 

> That is, when a package version is downloaded for the first time, SwiftPM trusts that 

Reviewer: ```suggestion
That is, when a package version is downloaded for the first time, the package manager trusts that 
```

it has downloaded the correct contents and requires subsequent downloads of the same 
package version to have the same fingerprint. 
If the fingerprint changes, it might be an indicator that the package has been

> compromised and SwiftPM will either warn or return an error.

Reviewer: ```suggestion
compromised and the package manager either warns or returns an error.
Control the choice of warning or error using the build option `--resolver-fingerprint-checking`, which defaults to `strict`.
Set the value to `warn` to not error on a mismatched fingerprint.
```

Depending on where a package version is downloaded from, a different value is
used as its fingerprint:
                             
| Package Version Origin | Fingerprint |
| ---------------------- | ----------- |
| Git repository         | Git hash of the revision |
| Package registry       | Checksum of the source archive |

> All version fingerprints of a package are kept in a single file

Reviewer: ```suggestion
The package manager keeps version fingerprints for each package in a file
```

under the `~/.swiftpm/security/fingerprints` directory.

>   - For a Git repository package, the fingerprint filename takes the form of `{PACKAGE_NAME}-{REPOSITORY_URL_HASH}.json` (e.g., `LinkedList-5ddbcf15.json`).

Reviewer: ```suggestion
  - For a Git repository package, the fingerprint filename takes the form of `{PACKAGE_NAME}-{REPOSITORY_URL_HASH}.json` (such as `LinkedList-5ddbcf15.json`).
```

>   - For a registry package, the fingerprint filename takes the form of `{PACKAGE_ID}.json` (e.g., `mona.LinkedList.json`).

Reviewer: ```suggestion
  - For a registry package, the fingerprint filename takes the form of `{PACKAGE_ID}.json` (such as `mona.LinkedList.json`).
```

> 

Reviewer: ```suggestion
For packages retrieved from a registry, the package manager expects all registries to provide consistent fingerprints for packages they host.
If registries have conflicting fingerprints, package manager reports that as an error.
```

> ### Using fingerprints for TOFU

Reviewer: Everything from line 30 (heading included) is replicated content from above - I made some suggestions to cherry pick a few details in the content above, and I think everything below can be deleted. Give it a read through and verify though.

Author: After re-reading, this makes sense - the only extra bit I've taken from `Using fingerprints for TOFU` is the part about converting the TOFU error to a warning using the `swift build` option `--resolver-fingerprint-checking`, wherein I link out to the `swift build` CLI article. Otherwise, all the content is essentially the same!
